Add option for enable/disable enum members in docstring.

knarfS opened the pull request:
## Description

This adds the option to disable/enable the enum members listed in the docstring (default is enabled). This is useful, when the documentation tool is able to generate documentation for class variables, so the enum members / class variable aren't listed twice in the documentation. An other benefit of using the features of the documentation generator is, that the members are linkable in the documentation.

I'm using pdoc3 to generate the documentation of an embedded python module, but the above also applies partially to pydoc. No idea about sphinx, I wasn't able to generate the documentation programmatically with sphinx...

I'd be happy to add that option to the documentation, but need a hint to which section this would fit best. There is a chapter about sphinx, but that doesn't exactly match.

## Suggested changelog entry:

```
Add option for enable/disable enum members in docstring.
```

knarfS commented:
>     1. Should this be a global setting, or just a local flag, passed to the enum constructor? I'm not sure; other reviewers, what do you think? At any rate, I'd like some arguments pro or contra this approach.

I wanted to put this in the same place as the other docstring related options (`function_signatures`and `user_defined_docstrings`), but yeah, maybe the enum ctor would be a better place?

>     2. `m_base.attr("__doc__") = static_property(...)` isn't doing anything more than adding something to the docstring. The `if` statement can/should just go around this whole creation of the `static_property`, I think.

There is still the docstring of the enum class itself (`std::string(((PyTypeObject *) arg.ptr())->tp_doc)`), which shouldn't be affected by this option.

YannickJadoul commented:
> I wanted to put this in the same place as the other docstring related options (`function_signatures`and `user_defined_docstrings`), but yeah, maybe the enum ctor would be a better place?

Yeah, I'm not 100% convinced of either approach, but I thought it's worth bringing up).
The advantage of a global thing is that it's consistent across all `enums_`. The thing that makes me doubt this approach is whether this isn't too fine-grained to make a full global option out of it.
The disadvantage of a constructor argument is that C++ doesn't have keyword arguments or even keyword-only arguments... :-/

If there are no further complaints, I'm happy to keep the global option, though.

> There is still the docstring of the enum class itself (`std::string(((PyTypeObject *) arg.ptr())->tp_doc)`), which shouldn't be affected by this option.

I believe that if you don't set `__doc__`, asking for `__doc__` will/should give you the contents of `tp_doc`. I might of course be completely wrong. Could you please give this a try?

knarfS commented:
> I believe that if you don't set `__doc__`, asking for `__doc__` will/should give you the contents of `tp_doc`. I might of course be completely wrong. Could you please give this a try?

You are right, I've just tested this and it works!

YannickJadoul commented:
Thanks, that's already better, I think :-)

>  The disadvantage of a constructor argument is that C++ doesn't have keyword arguments or even keyword-only arguments... :-/

Just to add to future discussion (I'm still not saying this is preferred): if we want to get around this, another option is to come up with some custom tag like `py::enum_members_docstring(false)` or so.

knarfS commented:
This will also fix https://github.com/pybind/pybind11/issues/2275 when disabling the enum members docstring (aka not setting `__doc__`):

```c++
# include <pybind11/pybind11.h>

namespace py = pybind11;

enum Numbers {
    zero,
    one
}

PYBIND11_MODULE(numbers, m) {
    py::options options;
    options.disable_enum_members_docstring();

    py::enum_<Numbers>e(m, "Numbers");
    e
        .value("zero", zero)
        .value("one", one)
        .export_values();
    e.attr("__doc__") = "test";
}
```

YannickJadoul commented:
Thanks for checking that as well, @knarfS!

rwgk commented:
Looks good to me, based on the understanding that this PR adds an option with a default that leaves the existing behavior unchanged, unless the option is actually used. Is that correct?

The only thing missing are tests and possibly a couple lines in the documentation (I'd look for documentation for existing options as a starting point).

knarfS commented:
> Looks good to me, based on the understanding that this PR adds an option with a default that leaves the existing behavior unchanged, unless the option is actually used. Is that correct?

Yes, the current docstring behavior is untouched. I've added a test to secure that.

I've also added a section about the undocumented option `disable_user_defined_docstrings()`.

rwgk commented on docs/advanced/misc.rst:
```diff
@@ -324,6 +324,15 @@ The class ``options`` allows you to selectively suppress auto-generated signatur
     }
 
+pybind11 also appends all members of an enum to the resulting enum docstring,
+this default behaviour can be disabled by using the ``disable_enum_members_docstring()``
```
I think this will read slightly nicer as a new sentence:
```
... resulting enum docstring.
This default behavior ...
```

rwgk commented on docs/advanced/misc.rst:
```diff
@@ -324,6 +324,15 @@ The class ``options`` allows you to selectively suppress auto-generated signatur
+function of the ``options`` class.
+
+With ``disable_user_defined_docstrings()`` all user defined docstrings of
+``module_::def()``, ``class_::def()`` and ``enum_()`` are discarded, but the
```
I'd use "disabled" instead of "discarded" here.

rwgk reviewed (approved):
Nice, thank you!

knarfS commented:
Thank you @rwgk for reviewing and approving. I updated the PR with your suggestions.

rwgk commented:
Great, thanks! I'll wait a day or so to give @Skylion007 I chance to review, too.

Skylion007 commented on include/pybind11/pybind11.h:
```diff
@@ -1972,29 +1972,32 @@ struct enum_base {
+                        dict entries = arg.attr("__entries");
+                        if (((PyTypeObject *) arg.ptr())->tp_doc) {
+                            docstring
+                                += std::string(((PyTypeObject *) arg.ptr())->tp_doc) + "\n\n";
```
nit: reinterpret_cast ?

rwgk replied:
Applied in commit c2456d9d3b8b154035f77e0895f7d81fc1cb0ccd

Skylion007 commented on include/pybind11/pybind11.h:
```diff
@@ -1972,29 +1972,32 @@ struct enum_base {
+                            auto comment = kv.second[int_(1)];
+                            docstring += "\n\n  " + key;
+                            if (!comment.is_none()) {
+                                docstring += " : " + (std::string) pybind11::str(comment);
```
nit: static_cast or `.cast<std::string>`

rwgk replied:
Also applied in Applied in commit c2456d9d3b8b154035f77e0895f7d81fc1cb0ccd

I also changed the code to systematically use `+=`.

Skylion007 commented on include/pybind11/pybind11.h:
```diff
@@ -1972,29 +1972,32 @@ struct enum_base {
+                        for (auto kv : entries) {
+                            auto key = std::string(pybind11::str(kv.first));
+                            auto comment = kv.second[int_(1)];
+                            docstring += "\n\n  " + key;
```
```suggestion
                            docstring += "\n\n  ";
                            docstring += key;
```

Skylion007 commented:
> Great, thanks! I'll wait a day or so to give @Skylion007 I chance to review, too.

Sorry, missed this in my Github notifications.

rwgk commented:
I went ahead with a rebase & applying the review suggestions, to get this into the next smart_holder update.

Waiting for the CI.

rwgk commented:
The only one CI failure is a known flake (test_gil_scoped DEADLOCK), definitely unrelated to this RP.

knarfS commented:
Thanks @rwgk  for merging!

Not sure, but maybe https://github.com/pybind/pybind11/issues/2275 can be also closed now ([see test above](https://github.com/pybind/pybind11/pull/2768#issuecomment-755266595)).

rwgk commented:
> Thanks @rwgk for merging!
> 
> Not sure, but maybe #2275 can be also closed now ([see test above](https://github.com/pybind/pybind11/pull/2768#issuecomment-755266595)).

@AWhetter Is there a chance you could confirm & update or close #2275?

AWhetter commented:
This does indeed fix that issue!

rwgk commented:
Thanks @AWhetter!
